kotlinlib/src/mill/kotlinlib/ksp/KspModule.scala

```diff
@@ -0,0 +1,156 @@
+package mill.kotlinlib.ksp
+
+import mill._
+import mill.api.{PathRef, Result}
+import mill.define.Task
+import mill.kotlinlib.worker.api.{KotlinWorker, KotlinWorkerTarget}
+import mill.kotlinlib.{Dep, DepSyntax, KotlinModule}
+
+import java.io.File
+
+/**
+ * Sets up the kotlin compiler for using KSP (Kotlin Symbol Processing)
+ * by plugging in the symbol-processing and symbol-processing-api dependencies.
+ *
+ * Use of kotlin-compiler-embedded is also recommended (and thus enabled by default)
+ * to avoid any classpath conflicts between the compiler and user defined plugins!
+ */
+@mill.api.experimental
+trait KspModule extends KotlinModule { outer =>
+
+  /**
+   * The version of the symbol processing library to use.
+   *
+   * This is combined with the version of the kotlin compiler to pull the symbol processing
+   * plugins for the compiler. These dependencies are
+   *
+   * com.google.devtools.ksp:symbol-processing-api
+   * and
+   * com.google.devtools.ksp:symbol-processing
+   *
+   * For more info go to [[https://kotlinlang.org/docs/ksp-command-line.html]]
+   * @return
+   */
+  def kspVersion: T[String]
+
+  /**
+   * Mandatory plugins that are needed for KSP to work.
+   * For more info go to [[https://kotlinlang.org/docs/ksp-command-line.html]]
+   *
+   * @return
+   */
+  def kspPlugins: T[Agg[Dep]] = Task {
+    Agg(
+      ivy"com.google.devtools.ksp:symbol-processing-api:${kotlinVersion()}-${kspVersion()}",
+      ivy"com.google.devtools.ksp:symbol-processing:${kotlinVersion()}-${kspVersion()}"
+    )
+  }
+
+  override def generatedSources: T[Seq[PathRef]] = Task {
+    generateSourcesWithKSP() ++ super.generatedSources()
+  }
+
+  final def kspPluginsResolved: T[Agg[PathRef]] = Task {
+    defaultResolver().resolveDeps(kspPlugins())
+  }
+
+  /**
+   * The symbol processors to be used by the Kotlin compiler.
+   * Default is empty.
+   */
+  def kotlinSymbolProcessors: T[Agg[Dep]] = Task {
+    Agg.empty[Dep]
+  }
+
+  final def kotlinSymbolProcessorsResolved: T[Agg[PathRef]] = Task {
+    defaultResolver().resolveDeps(
+      kotlinSymbolProcessors()
+    )
+  }
+
+  override def kotlinCompilerEmbeddable: Task[Boolean] = Task { true }
+
+  /*
+   * The symbol processing plugin id
+   */
+  private val kspPluginId: String =
+    "com.google.devtools.ksp.symbol-processing"
+
+  /*
+   * The Kotlin compile task with KSP.
+   * This task should run as part of the [[generatedSources]] task to
+   * so that the generated  sources are in the [[compileClasspath]]
+   * for the main compile task.
+   */
+  private def generateSourcesWithKSP = Task {
```

Comment:
Since generating files is the whole point of this module, this task should be `public`.